README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

A flutter client library for the [IPFS HTTP API](https://docs.ipfs.io/reference/http/api/).

### Features

* Make a directory
* List all directories/files in a directory/subdirectory
* File status
* Write a file
* Read a file
* Remove a file/directory
* Move/Rename a file

### Usage
```
  IpfsClient ipfsClient = IpfsClient();
```
OR
```
  IpfsClient ipfsClient = IpfsClient(url: "http://127.0.0.1:5001");
  // default is http://127.0.0.1:5001 so you don't need to pass url in case you are working on localhost
```

If your app uses `Basic Authorization` then,
```
  IpfsClient ipfsClient =
      IpfsClient(url: "YOUR_SERVER_URL", authorizationToken: "YOUR_TOKEN");
```
### How to make a directory in IPFS
```
  var response = await ipfsClient.mkdir(dir: 'myfolder');
```

### Get list of directories/files in a directory/subdirectory in the local mutable namespace.
```
  var response = await ipfsClient.ls(dir: 'myfolder');
```
If you dont pass any folder it will return all files and directories of root directory.

### How to Write to a file
```
    var response = await ipfsClient.write(
      dir: 'myfolder/sample.png',
      filePath: "/Users/myuser/Desktop/sample.png",
      fileName: "sample.png");
```
### How to read a file
```
  var response = await ipfsClient.read(dir: "myfolder/sample.png");
```
### How to remove a file
```
  var response = await ipfsClient.remove(dir: "myfolder/sample.png");
```
### How to move/rename a file
```
  var response = await ipfsClient.mv(oldPath: "myfolder/sample.png", newPath: "myfolder/MY_SAMPLE.png");
```


CHANGELOG.md
## 1.0.5

- recursive: Recursively remove directories


test/ipfs_client_flutter_test.dart
import 'package:ipfs_client_flutter/ipfs_client_flutter.dart';
import 'package:test/test.dart';

void main() {
  // group('A group of tests', () {
  //   final awesome = Awesome();
  //
  //   setUp(() {
  //     // Additional setup goes here.
  //   });
  //
  //   test('First Test', () {
  //     expect(awesome.isAwesome, isTrue);
  //   });
  // });
}


lib/src/ipfs_service.dart
// TODO: Put public facing types in this file.
import 'package:dio/dio.dart';

/// Checks if you are awesome. Spoiler: you are.
class IpfsService {
  Dio dio = Dio();

  IpfsService() {
    dio.interceptors.add(LogInterceptor());
    //dio.interceptors.add(LogInterceptor(requestBody: true));
    // (dio.httpClientAdapter as DefaultHttpClientAdapter).onHttpClientCreate =
    //     (HttpClient client) {
    //   client.findProxy = (uri) {
    //     //proxy all request to localhost:8888
    //     return 'PROXY localhost:8888';
    //   };
    //   client.badCertificateCallback =
    //       (X509Certificate cert, String host, int port) => true;
    // };
  }

  Future<dynamic> postFile(
      {required String url,
      var queryParameters,
      var formData,
      var authorizationToken}) async {
    try {
      Response response = await dio.post(
        url,
        data: formData,
        queryParameters: queryParameters,
        options: Options(
          headers: {
            "Authorization": "Basic $authorizationToken",
          },
        ),
        onSendProgress: (received, total) {
          if (total != -1) {
            print((received / total * 100).toStringAsFixed(0) + '%');
          }
        },
      );
      return response.data;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  Future<dynamic> post(
      {String? url,
      var queryParameters,
      var authorizationToken,
      ResponseType responseType = ResponseType.plain}) async {
    try {
      Response response = await dio.post(url!,
          queryParameters: queryParameters,
          options: Options(
            responseType: responseType,
            headers: {
              "Authorization": "Basic $authorizationToken",
            },
          ));
      return {"statusCode": response.statusCode, "data": response.data};
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }
}


lib/src/ipfs_client_flutter_base.dart
// TODO: Put public facing types in this file.

import 'package:dio/dio.dart';
import 'ipfs_service.dart';

/// Checks if you are awesome. Spoiler: you are.
class IpfsClient {
  String? url;
  String? authorizationToken;
  final IpfsService _ipfsService = IpfsService();

  IpfsClient({this.url = 'http://127.0.0.1:5001', this.authorizationToken});

  /// Make a directory in IPFS
  /// For more: https://docs.ipfs.io/reference/http/api/#api-v0-files-mkdir
  Future<dynamic> mkdir({required String dir}) async {
    _checkDir(dir);
    var params = {
      'arg': dir,
      'parent': true,
    };
    try {
      var response = await _ipfsService.post(
          url: '$url/api/v0/files/mkdir?',
          queryParameters: params,
          authorizationToken: authorizationToken);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  /// List directories in the local mutable namespace.
  /// For more: https://docs.ipfs.io/reference/http/api/#api-v0-files-ls
  Future<dynamic> ls({required String dir}) async {
    _checkDir(dir);
    var params = {
      'arg': dir,
      'long': true,
      'U': true,
    };
    try {
      var response = await _ipfsService.post(
          url: '$url/api/v0/files/ls?',
          queryParameters: params,
          authorizationToken: authorizationToken);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  /// Display file status.
  /// For more: https://docs.ipfs.io/reference/http/api/#api-v0-files-stat
  Future<dynamic> stat(
      {required String dir, Map<String, dynamic> params = const {}}) async {
    _checkDir(dir);
    var _params = {'arg': dir, ...params};
    try {
      var response = await _ipfsService.post(
          url: '$url/api/v0/files/stat?',
          queryParameters: _params,
          authorizationToken: authorizationToken);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  /// Write to a file in a given filesystem.
  /// For more: https://docs.ipfs.io/reference/http/api/#api-v0-files-write
  Future<dynamic> write(
      {required String dir,
      required String filePath,
      String fileName = ""}) async {
    _checkDir(dir);
    var params = {
      'arg': dir,
      'create': true,
    };
    try {
      var response = await _ipfsService.postFile(
          url: '$url/api/v0/files/write?',
          formData: await _getFormData(filePath, fileName),
          queryParameters: params,
          authorizationToken: authorizationToken);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  /// Read a file in a given MFS.
  /// For more: https://docs.ipfs.io/reference/http/api/#api-v0-files-read
  Future<dynamic> read({required String dir}) async {
    _checkDir(dir);
    var params = {
      'arg': dir,
      'offset': 0,
      'count': 10000000,
    };
    try {
      var response = await _ipfsService.post(
          url: '$url/api/v0/files/read?',
          queryParameters: params,
          authorizationToken: authorizationToken,
          responseType: ResponseType.bytes);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  /// Remove a file/dir
  /// https://docs.ipfs.io/reference/http/api/#api-v0-files-rm
  /// recursive [bool]: recursively remove directories. Required: no.
  Future<dynamic> remove({required String dir, bool recursive = true}) async {
    _checkDir(dir);
    var params = {
      'arg': dir,
      'recursive': recursive,
    };
    try {
      var response = await _ipfsService.post(
          url: '$url/api/v0/files/rm?',
          queryParameters: params,
          authorizationToken: authorizationToken);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  /// move a file/mv
  /// https://docs.ipfs.io/reference/http/api/#api-v0-files-mv
  Future<dynamic> mv({required String oldPath, required String newPath}) async {
    _checkDir(oldPath);
    _checkDir(newPath);
    try {
      var response = await _ipfsService.post(
          url: '$url/api/v0/files/mv?arg=$oldPath&arg=$newPath',
          // queryParameters: params,
          authorizationToken: authorizationToken);
      return response;
    } on DioError catch (e) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and is also not 304.
      if (e.response != null) {
        return e.response?.data;
      }
    }
  }

  String _checkDir(String dir) {
    if (!dir.startsWith("/")) return dir = "/$dir";
    return dir;
  }

  Future<FormData> _getFormData(String filePath, String fileName) async {
    return FormData.fromMap({
      'file': await MultipartFile.fromFile(
        filePath,
        filename: fileName,
      ),
    });
  }
}
